### PR TITLE
Optimize llm/GPT3 performance

### PR DESCRIPTION
**Pull request by MarioLulab**
### PR types
Performance optimization

### PR changes
Others

### Description
将 model_zoo/gpt-3 中的优化策略迁移至 llm/gpt-3 中

- [x] Fast Layer Norm op
开启方法：运行 `run_pretrain.py` 时设置 `--use_fast_layer_norm true`

- [x] Fused Linear
开启方法：运行 `run_pretrain.py` 时设置 `--use_fused_linear true`

- [x] Fused Dropout + Add Residual
开启方法：运行 `run_pretrain.py` 时设置 `--use_fused_dropout_add true`

- [x] enable_linear_fused_grad_add
开启方法：版本要求：paddlenlp >= a5d87f5。运行 `run_pretrain.py` 时设置 `--enable_linear_fused_grad_add true`

- [x] use SPInnerOverlap
开启方法：版本要求：paddlenlp >= a092775, paddle >= cfaa001，运行 `run_pretrain.py` 时设置 `--tensor_parallel_config enable_mp_async_allreduce enable_fused_linear_param_grad_add ` 并且训练开启 sp

- [x] disable transmission attention_mask while pp
开启方法：默认在 pp 时不传输 attention_mask

### CODE REVIEW

File: paddlenlp/transformers/gpt/modeling_pp.py [hunk 1]
```diff
@@ -103,15 +103,13 @@ def forward(self, args):
...
-        causal_mask = self.bias[:, :, 0:seq_length, :seq_length]
         if attention_mask is not None:
             if attention_mask.dtype != paddle.int64:
                 attention_mask = paddle.cast(attention_mask, dtype=paddle.int64)
             if len(attention_mask.shape) == 2:
                 attention_mask = attention_mask[:, None, None, :]
+            causal_mask = self.bias[:, :, 0:seq_length, :seq_length]
             attention_mask = (1.0 - (attention_mask & causal_mask)) * -1e4
-        else:
-            attention_mask = (1.0 - causal_mask) * -1e4
```
Comment (Xreki) [inline paddlenlp/transformers/gpt/modeling_pp.py:114]: 原来`else`分支是生成了一个attention_mask吧，不用Flash-Attention的时候是需要用到这个attention_mask的？

File: tests/transformers/gpt/test_modeling.py [hunk 2]
```diff
@@ -369,6 +369,25 @@ def create_and_check_gpt_for_token_classification(
         else:
             self.parent.assertEqual(result.shape, [self.batch_size, self.seq_length, self.num_labels])
 
+    def create_and_check_gpt_for_fused_linear_and_fused_dropout_add(self, config, input_ids, *args):
```
Comment (Xreki) [inline tests/transformers/gpt/test_modeling.py]: CI如果是在CPU上跑的话，这个单测没有必要加了

File: paddlenlp/transformers/gpt/modeling_pp.py (hunk 1 above)
Comment (MarioLulab) [reply paddlenlp/transformers/gpt/modeling_pp.py:114]: 我的理解是也不需要。在不使用 Flash-Attention 时，会使用 _core_attention，其会在 attention_mask = None 时使用 `incubate.softmax_mask_fuse_upper_triangle` 达到等价的效果: https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/gpt/modeling.py#L338

image

如此可以避免 attention_mask 的反复传输带来的额外开销了

File: tests/transformers/gpt/test_modeling.py (hunk 2 above)
Comment (MarioLulab) [reply tests/transformers/gpt/test_modeling.py]: 好的，将删除添加的单测

File: llm/run_pretrain.py [hunk 3]
```diff
@@ -154,6 +155,18 @@ class ModelArguments:
         default=False,
         metadata={"help": "llama or other model, use_fused_rms_norm"},
     )
+    use_fast_layer_norm: bool = field(
+        default=False,
+        metadata={"help": "GPT3 model, use fast layernorm"},
+    )
```
Comment (ZHUI) [inline llm/run_pretrain.py:161]: 有收益吗？我记得 paddle layer norm已经很快了

Comment (MarioLulab) [reply llm/run_pretrain.py:161]: 使用 GPT3-13B，在mp2-pp4的配置下预训练测试，实验结果是：开启fast_ln 会比baseline 提升1.07%的训练吞吐

**Review by ZHUI — APPROVED** [review]
LGTM
